libcfnet/protocol.c

```diff
@@ -88,3 +89,93 @@ Seq *ProtocolOpenDir(AgentConnection *conn, const char *path)
 
     return seq;
 }
+
+int ProtocolStat(AgentConnection *conn, const char *remote_path,
+                 struct stat *stat_buf)
+{
+    assert(conn != NULL);
+    assert(remote_path != NULL);
+    assert(stat_buf != NULL);
+
+    time_t tloc = time(NULL);
+    if (tloc == (time_t) -1)
+    {
+        Log(LOG_LEVEL_ERR, "Couldn't read system clock (time: %s)",
+            GetErrorStr());
+        tloc = 0;
+    }
+
+    char buf[CF_BUFSIZE] = {0};
+    // Won't this cause problems on 32-bit systems? time_t would always be 64-bit, right?
+    int to_send = snprintf(buf, CF_BUFSIZE, "SYNCH %jd STAT %s",
+                           (intmax_t) tloc, remote_path);
+
+    int ret = SendTransaction(conn->conn_info, buf, to_send, CF_DONE) == -1;
+    if (ret == -1)
+    {
+        Log(LOG_LEVEL_ERR,
+            "%s: Could not send stat request for file %s:%s.",
+            __func__, conn->this_server, remote_path);
+        return -1;
+    }
+
+    int recvd_len = ReceiveTransaction(conn->conn_info, buf, NULL) == -1;
+    if (recvd_len == -1)
+    {
+        Log(LOG_LEVEL_ERR,
+            "%s: Receiving from remote host %s failed!",
+            __func__, conn->this_server);
+        return -1;
+    }
+
+    if (BadProtoReply(buf))
+    {
+        Log(LOG_LEVEL_ERR,
+            "%s: Server thought this was bad: %s",
+            __func__, buf);
+        return -1;
+    }
+
+    if (!OKProtoReply(buf))
+    {
+        Log(LOG_LEVEL_ERR,
+            "%s: Illegal response from server",
+            __func__);
+        return -1;
+    }
+
+    Stat cf_stat;
+    ret = StatParseResponse(buf, &cf_stat);
+    if (ret == -1)
+    {
+        Log(LOG_LEVEL_ERR,
+            "%s: Failed to parse the response from the server",
+            __func__);
+        return -1;
```

Comment:
I don't think printing function names in `ERR` messages which are shown to the user by default is a good idea. It's good for `DEBUG` messages, but I think it only adds confusion to `ERR` messages.